massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+            };
+        }
+
+        if let Some(encoding) = &config.accept_compressed {
+            if encoding.eq_ignore_ascii_case("Gzip") {
+                svc = svc.send_compressed(CompressionEncoding::Gzip);
+            };
+        }
+
+        let (shutdown_send, shutdown_recv) = oneshot::channel::<()>();
+
+        if config.accept_http1 {
+            let mut router_with_http1 = tonic::transport::Server::builder()
+                .accept_http1(true)
+                .layer(GrpcWebLayer::new())
+                .add_service(svc);
+
+            if config.enable_reflection {
+                let reflection_service = tonic_reflection::server::Builder::configure()
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router_with_http1 = router_with_http1.add_service(reflection_service);
+            }
+            router_with_http1
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        } else {
+            let mut router = tonic::transport::Server::builder().add_service(svc.clone());
+
+            if config.enable_reflection {
+                let reflection_service = tonic_reflection::server::Builder::configure()
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router = router.add_service(reflection_service);
+            }
+
+            router
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        }
+
+        Ok(StopHandle {
+            server_handler: shutdown_send,
+        })
+    }
+}
+
+/// Used to be able to stop the gRPC API
+pub struct StopHandle {
+    server_handler: oneshot::Sender<()>,
+}
+
+impl StopHandle {
+    /// stop the gRPC API gracefully
+    pub fn stop(self) {
+        match self.server_handler.send(()) {
+            Ok(_) => {
+                info!("gRPC API finished cleanly");
+            }
+            Err(err) => warn!("gRPC API thread panicked: {:?}", err),
+        }
+    }
+}
+
+fn match_for_io_error(err_status: &tonic::Status) -> Option<&std::io::Error> {
+    let mut err: &(dyn Error + 'static) = err_status;
+
+    loop {
+        if let Some(io_err) = err.downcast_ref::<std::io::Error>() {
+            return Some(io_err);
+        }
+
+        // h2::Error do not expose std::io::Error with `source()`
+        // https://github.com/hyperium/h2/pull/462
+        if let Some(h2_err) = err.downcast_ref::<h2::Error>() {
+            if let Some(io_err) = h2_err.get_io() {
+                return Some(io_err);
+            }
+        }
+
+        err = match err.source() {
+            Some(err) => err,
+            None => return None,
+        };
+    }
+}
+
+#[tonic::async_trait]
+impl grpc::grpc_server::Grpc for MassaService {
+    async fn get_selector_draws(
```

Comment:
if possible we need to improve all function definition 

```rust
    async fn get_selector_draws(
        &self,
        request: tonic::Request<grpc::GetSelectorDrawsRequest>,
    ) -> Result<tonic::Response<grpc::GetSelectorDrawsResponse>, tonic::Status>
```
to something like this 

```rust
    async fn get_selector_draws(
        &self,
        request: GetSelectorDrawsRequest,
    ) -> Result<GetSelectorDrawsResponse, MassaGrpcError>
```

Reply:
Yeah, but for namespace, we'll got a conflict. As the Protobuf structures have the same name as existing one. THis was saw we Sylvain, may be we'll write a proc macro to auto generated the bindings between proto::Slot and massa::Slot for example

Reply:
oh yes ok ! we will do it later

Reply:
I saw different options, IMHO we should keep `tonic::Status` as top level, but implement From for our custom errors. As we are autogenerating the code, we can't change the signatures of functions in generated Trait.